bokeh/models/sources.py

```diff
@@ -757,3 +789,23 @@ class AjaxDataSource(RemoteSource):
         ajax_source.headers = { 'x-my-custom-header': 'some value' }
 
     """)
+
+#-----------------------------------------------------------------------------
+# Dev API
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Private API
+#-----------------------------------------------------------------------------
+
+def _check_slice(s):
+    if (s.start is not None and s.stop is not None and s.start > s.stop):
+        raise ValueError("Patch slices must have start < end, got %s" % s)
+    if (s.start is not None and s.start < 0) or \
+       (s.stop  is not None and s.stop < 0) or \
+       (s.step  is not None and s.step < 0):
+        raise ValueError("Patch slices must have non-negative (start, stop, step) values, got %s" % s)
```

Comment:
@HuntJSparra I think the meaning of the sections should be clarified for test file. A section "Private API" here means *tests for private API functions*, not just a place for underscore functions in the test module itself. (In fact, I'm not sure why anything should be underscore in a test module since users don't interact with them anyway, but that's a separate question) I think helper functions like this in test files should go at the top in the "setup" section, or else just nearby where they are used

Reply:
bokeh/models/sources.py is not a test module though, correct?

Reply:
It's not, my apologies, obviously trying to do too many things at once this morning...